system/security/plugins/testauthSecurity/testauthSecurity.cpp

```diff
@@ -0,0 +1,572 @@
+    SecAccessFlags defaultFeaturAccess = SecAccess_Unavailable;
+    MapStringTo<int> featureAccesses, eclwuScopeAccesses;
+    Owned<IPropertyTree> fileScopeAccesses;
+
+public:
+    CUserAccess(const char* _userName, const char* _password)
+        : userName(_userName), password(_password)
+    {
+        fileScopeAccesses.setown(createPTree("filescopes"));
+    };
+
+    const char* queryUserName()
+    {
+        return userName.get();
+    }
+
+    const char* queryPassword()
+    {
+        return password.get();
+    }
+
+    void setDefaultFeatureAccess(const char* access)
+    {
+        defaultFeaturAccess = getSecAccessFlagValue(access);
+    }
+
+    void setDefaultFileScopeAccess(const char* access)
+    {
+        defaultFileScopeAccess = getSecAccessFlagValue(access);
+    }
+
+    void setDefaultECLWUScopeAccess(const char* access)
+    {
+        defaulteECLWUScopeAccess = getSecAccessFlagValue(access);
+    }
+
+    void addFeatureAccess(const char* resource, const char* access)
+    {
+        SecAccessFlags flag = getSecAccessFlagValue(access);
+        if (!featureAccesses.getValue(resource))
+            featureAccesses.setValue(resource, flag);
+    }
+
+    void addECLWUScopeAccess(const char* scope, const char* access)
+    {
+        SecAccessFlags flag = getSecAccessFlagValue(access);
+        if (!eclwuScopeAccesses.getValue(scope))
+            eclwuScopeAccesses.setValue(scope, flag);
+    }
+
+    void addFileScopeAccess(const char* scope, const char* access)
+    {
+        //Convert the scope to xpath.
+        StringArray scopes;
+        if (!convertFileScopeToXPath(scope, true, scopes))
+        {
+            OWARNLOG("Invalid scope %s.", scope);
+            return;
+        }
+
+        //Check duplicate
+        const char* xpath = scopes.item(0);
+        if (fileScopeAccesses->hasProp(xpath))
+        {
+            OWARNLOG("Duplicated accesss settting for scope %s.", scope);
+            return;
+        }
+
+        //Add a tree node for the access.
+        IPropertyTree* accessNode = ensurePTree(fileScopeAccesses, xpath);
+        accessNode->setPropInt("@access", getSecAccessFlagValue(access));
+    }
+
+    SecAccessFlags queryFeatureAccess(const char* resource)
+    {
+        int* allow = featureAccesses.getValue(resource);
+        return allow ? (SecAccessFlags) *allow : defaultFeaturAccess;
+    }
+
+    /* same logic as in LDAP security
+    if perms set on 'scopeA::scopeB' only and lookup of 'scopeA::scopeB::scopeC::scopeD'
+    need to lookup:
+        'scopeA'
+    no match=>continue
+    match=>continue if read permissions (if no read, implies can't "see" child scopes)
+        'scopeA::scopeB'
+    no match=>continue
+    match=>continue if read permissions (if no read, implies can't "see" child scopes)
+    etc. Until full scope path checked, or no read permissions hit on ancestor scope.
+    */
+    SecAccessFlags queryFileScopeAccess(const char* scope)
```

Comment:
I was picturing something a little more like the following (untested pseudo code):

```
SecAccessFlags getFileScopeAccess(IPropertyTree *accessTree, const StringArray &nodes, unsigned pos, SecAccessFlags access, bool first)
{
    if (!nodes.isItem(pos))
        return access;
    const char *node = nodes->item(pos);
    if (isEmptyString(node))
        return access;
    IPropertyTree *element = accessTree->queryPropTree(node);
    if (!element)
        return access;
    if (element->hasProp("@access"))
    {
        int newAccess = element->queryPropInt("@access);
        if (first || newAccess < access) //least access wins
        {
            access = newAccess;
            first = false;
        }
    }
    return getFileScopeAccess(element, nodes, ++pos, access, first);
}

SecAccessFlags queryFileScopeAccess(const char* scope)
{
    StringArray nodes;
    if (!convertFileScopeToNodes(scope, false, nodes))  //==>  a::b::c::d becomes nodes[a, b, c, d]
        return SecAccess_Unavailable;
    return getFileScopeAccess(fileScopeAccesses, nodes, 0, defaultFileScopeAccess, true);
}
```

Basically if no @access is found it should use the default, but as it hits each node that has an '@access' it will return the lowest found.  (I.e. if a parent is READ only you can't have a child with FULL access.)

Reply:
Fixed